### DISCUSSION TITLE
How to export organised files into folders named by geolocation and date?

### DISCUSSION THREAD

**Discussion by sgarratt77** [RhetTbull/osxphotos#1484] [Q&A]
Hi Rhet,

I'm impressed that you continue to develop your script, seeing templates added now - nice!

I wanted to export any further libraries into folders that are named by geolocation firstly and then subfolders that have date (year and month). AS I type this I think it would be beneficial in also naming the folders by album name instead of geolocation. However not all media will have been put into an album and may be in multiple albums so it presents a problem. the best way I keep things organised in photos app is to crate albums entitled by 'event', ie; [2024-03-02] Trip to Park. I also use the @ symbol at times to help find locations I have entitled the album with too - not sure if this causes a problem with your script or not???

My question is, is it possible to achieve this and could you give me an example please?

My current script is this (now a year old) so I will need to amend it:
osxphotos export /Volumes/T7/Temp/Temp/Photo\ LIbraries\ to\ Export/1/Media --db /Volumes/T7/Temp/Temp/Photo\ LIbraries\ to\ Export/1.photoslibrary --exportdb /Volumes/T7/Temp/Temp/Photo\ LIbraries\ to\ Export/1/.osxphotos_export.db --export-by-date --touch-file --preview-if-missing --update --exiftool-merge-keywords --exiftool-merge-persons --album-keyword --person-keyword --strip --sidecar xmp --report /Volumes/T7/Temp/Temp/Photo\ LIbraries\ to\ Export/1/Report1Offline.csv --save-config /Volumes/T7/Temp/Temp/Photo\ LIbraries\ to\ Export/1/Config1Offline.toml --keyword-template "PhotosExport>{folder_album(>)}" --keyword-template "{edited?edited-before-lightroom,}" --exiftool 2>1.log

**Comment by RhetTbull** [comment]
>  I also use the @ symbol at times to help find locations I have entitled the album with too - not sure if this causes a problem with your script or not???

That's no problem. OSXPhotos will sanitize any folder or file name to ensure it is compatible with the file system. `@` works fine.

> I wanted to export any further libraries into folders that are named by geolocation firstly and then subfolders that have date (year and month).

Remove `--export-by-date` and add `--directory "{place.name}/{created.year}/{created.mm}"`  This will output a folder structure like this:

`Washington, District of Columbia, United States/2024/03`

Some photos may not have reverse geolocation data so in that case, the template system would use the default value of `_` so you'd get: `_/2024/03`.  You can specify your own default value like this: `--directory "{place.name,NoPlace}/{created.year}/{created.mm}"`  in which case you'd get: `NoPlace/2024/03`.

There is a lot of reverse geolocation data available. See [here](https://rhettbull.github.io/osxphotos/template_help.html) and search for `{place`.  For example, `{place.name.city}` would list just the city.

> AS I type this I think it would be beneficial in also naming the folders by album name instead of geolocation.

You can do that with `{album}/{created.year}/{created.mm}` or `{folder_album}/{created.year}/{created.mm}` if you want the parent folders included.  Same pattern to specify a default value: `{album,NoAlbum}/{created.year}/{created.mm}`.

If a photo is in multiple albums you will get multiple copies of the photo exported. There is no way to avoid that.

**Reply by sgarratt77** [reply]
oh wow! That's really helpful, thanks Rhet!

What do you mean by "parent folders included"? like if my albums are within folders on photos app? i could see that being useful too if so. thanks

**Reply by RhetTbull** [reply]
"parent folders" refers to albums that are in folders in Photos. For example, in Photos, you have `Folder1>SubFolder1>Album1`, then `{folder_album}` gives you `Folder1/SubFolder1/Album1` during export.

**Comment by sgarratt77** [comment]
Also, can I also run my previously exported files/folders through using an adapted script so that I can update them to the new, desired structure as well? or is there a way I could do that? (maybe just using ExifTool on its own if osxphotos can only be used with existing photo library files?)

**Reply by RhetTbull** [reply]
Unfortunately there is no way to do this. If you change the output directory structure, you'd need to basically start over with a fresh export. Even if you used an external tool or script to rewrite the directory structure, OSXPhotos wouldn't know which photos the new folders refer to.  OSXPhotos maintains an export database in the export folder (`.osxphotos_export.db`) that it uses to maintain state and map each exported file path to a specific asset in the Photos library. If you change the path structure, the database would no longer map the paths correctly. You really need to decide on the export structure you'd like before you start backing up with OSXPhotos.

**Reply by sgarratt77** [reply]
ok. understood. what about importing the folders back into a new photo library as REFERENCED files?

**Reply by RhetTbull** [reply]
You should be able to do that. I strongly recommend against using referenced files though. The AI/ML features of Photos don't work well with referenced files and if you ever move to a new Mac or move the referenced files to a new drive Photos won't be able to find the files due to the fact that it uses sandboxes security scoped bookmarks to track the files. Referenced libraries are basically broken since Catalina.

**Reply by sgarratt77** [reply]
i wouldn't be worried about tracking the files and folders once I had ran the organising operation. i'd just use it to organise the files into well-named folders and that would be it. in light of this would you still advise not to do it as referenced?

if not do you know of another app or script that you have encountered that would allow me to do this?

**Reply by RhetTbull** [reply]
Photos won't organize the files on disk. I'm not sure exactly what you're looking for. Are you looking for an app to let you organize your photos (and adjust where they are stored on disk to match the organization)? If so, I don't use a workflow like this and am not familiar with any apps that do this. Maybe something like DigiKam or Lightroom.

**Reply by sgarratt77** [reply]
yes I'm trying to organise my existing (already exported) photos and videos into folders with the above-mentioned event names and include month and year.

**Reply by RhetTbull** [reply]
Why not just re-export with the updated folder structure?  I'm not aware of any tools to do this though I'm sure such exists. It's just not something I use so I'm not familiar with what's out there.

**Reply by sgarratt77** [reply]
i would re-export but i don't have the photo library files any longer. My ultimate aim is to organise my media into folders and use them within synology photos on my nas. i want to move away from apple photos completely really.

**Reply by sgarratt77** [reply]
ill look into other apps and can mention any good ones here in case it helps anyone else if you like?
